torchtune/modules/vision_transformer.py

```diff
@@ -0,0 +1,334 @@
+            >>>
+            >>> # (bsz, num_tiles, num_patches_per_tile + CLS token, embed_dim)
+            >>> print(x.shape)
+            torch.Size([1, 2, 101, 32]
+            >>>
+            >>> # (bsz, len(indices_return_hidden), num_tiles, num_patches_per_tile + CLS token, embed_dim)
+            >>> print(hidden_states.shape)
+            torch.Size([1, 5, 2, 101, 32]
+        """
+        hidden_states = []
+
+        # parse inputs
+        if images.ndim == 4:
+            n_tiles = 1
+            bsz, nch, w, h = images.shape
+            assert (
+                aspect_ratio is None or aspect_ratio.max() == 1
+            ), f"since images.ndim == 4, aspect ratio must be none or 1x1, got {aspect_ratio}"
+
+        elif images.ndim == 5:
+            bsz, n_tiles, nch, w, h = images.shape
+
+        else:
+            raise ValueError(
+                f"Unsupported number of dimensions: {images.ndim}. Expected 4 or 5."
+            )
+
+        # if aspect_ratio is not provided, it defaults to one tile [1,1]
+        if aspect_ratio is None:
+            aspect_ratio = torch.tensor([[1, 1]] * bsz, device=images.device)
+
+        images = images.reshape(bsz * n_tiles, nch, w, h)
+        aspect_ratio = aspect_ratio.reshape(bsz, 2)
+
+        # patch embeddings (tokens)
+        # out: (bsz, embed_dim, patch_grid_size, patch_grid_size)
+        x = self.conv(images)
+
+        # out: (bsz, patch_grid_size**2, embed_dim)
+        x = x.flatten(start_dim=2).permute(0, 2, 1)
+
+        _, n_tokens, embed_dim = x.shape
+
+        # pre_tile_pos_embed
+        if self.pre_tile_pos_embed:
+            x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+            x = self.pre_tile_pos_embed(x, aspect_ratio)
+
+        # apply cls token
+        x = x.reshape(bsz * n_tiles, n_tokens, embed_dim)
+        x = self.cls_token_embedding(x)
+        n_tokens += 1
+
+        # token_pos_embedding
+        x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+        x = self.token_pos_embedding(x, aspect_ratio)
+
+        # norm
+        x = self.ln_pre(x)
+
+        # transformer with optional hidden layer outputs
+        x = x.view(bsz, n_tiles * n_tokens, embed_dim)
+        for layer_idx, transformer_layer in enumerate(self.transformer_layers):
+            if self.indices_return_hidden and layer_idx in self.indices_return_hidden:
+                hidden_states.append(x)
+            x = transformer_layer(x)
+
+        # norm
+        x = self.ln_post(x)
+
+        # reshape outputs
+        x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+
+        if self.indices_return_hidden:
+            hidden_states = torch.stack(hidden_states, dim=-1)
+            hidden_states = hidden_states.reshape(
+                bsz, n_tiles, n_tokens, embed_dim, len(self.indices_return_hidden)
+            )
+        else:
+            hidden_states = torch.empty(0)  # dummy tensor
```

Comment:
The issue is that hidden_states is originally a list. Then, we replace it with a torch.stack or torch.empty.

If we add hidden_states = torch.empty(0) right before the for-loop, you can NOT run the next line anymore:
``hidden_states = torch.stack(hidden_states, dim=-1)``

If we want to enforce the pattern you suggested above, we would have to rename hidden_states to something else, like:

```
list_hidden_states = []
[...]
hidden_states = torch.empty(0)
if self.indices_return_hidden:
    hidden_states = torch.stack(list_hidden_states, dim=-1)
```

Btw, I wonder if it would be meaningful to pre allocating memory for the hidden states, instead of appending to list + stack.

Reply:
ok yeh that makes sense, totally zoinked on the fact that hidden_states is built up already. Sorry about that - perils of reviewing too quickly